### ISSUE TITLE
p11tool --list-all outputs ERROR: Esys_StartAuthSession: 0x902

### ISSUE THREAD

**Issue by dnoliver** [tpm2-software/tpm2-pkcs11#394]
Versions:

1. tpm2-tss@2.3.0
2. tpm2-tools@4.0.1
3. tpm2-pkcs11@1.0

The following test script

```bash
#!/bin/bash

set -euxo pipefail

export TPM2TOOLS_TCTI="device:/dev/tpmrm0"

tpm2_clear
rm tpm2_pkcs11.sqlite3
tpm2_ptool init
tpm2_ptool addtoken --pid=1 --sopin=mysopin --userpin=myuserpin --label=label
tpm2_ptool addkey --algorithm=rsa2048 --label=label --userpin=myuserpin
p11-kit list-modules
TOKEN=$(p11tool --list-token-urls | grep Intel)
expect <(cat <<EOF
spawn p11tool --login --list-all "${TOKEN}"
expect "Enter PIN: "
send -- "myuserpin\r"
interact
EOF
)
```

Outputs an error in the terminal (but p11tool ends with exit code 0)

```bash
[root@localhost ~]# bash init-pkcs11-store.sh 
+ export TPM2TOOLS_TCTI=device:/dev/tpmrm0
+ TPM2TOOLS_TCTI=device:/dev/tpmrm0
+ tpm2_clear
+ rm tpm2_pkcs11.sqlite3
+ tpm2_ptool init
action: Created
id: 1

+ tpm2_ptool addtoken --pid=1 --sopin=mysopin --userpin=myuserpin --label=label
+ tpm2_ptool addkey --algorithm=rsa2048 --label=label --userpin=myuserpin
action: add
private:
  CKA_ID: '33386436326437653365646138346562'
public:
  CKA_ID: '33386436326437653365646138346562'
+ p11-kit list-modules
p11-kit-trust: p11-kit-trust.so
    library-description: PKCS#11 Kit Trust Module
    library-manufacturer: PKCS#11 Kit
    library-version: 0.23
    token: System Trust
        manufacturer: PKCS#11 Kit
        model: p11-kit-trust
        serial-number: 1
        hardware-version: 0.23
        flags:
               token-initialized
    token: Default Trust
        manufacturer: PKCS#11 Kit
        model: p11-kit-trust
        serial-number: 1
        hardware-version: 0.23
        flags:
               write-protected
               token-initialized
opensc: opensc-pkcs11.so
    library-description: OpenSC smartcard framework
    library-manufacturer: OpenSC Project
    library-version: 0.19
tpm2_pkcs11: libtpm2_pkcs11.so
    library-description: TPM2.0 Cryptoki
    library-manufacturer: tpm2-software.github.io
    library-version: 1.0
    token: label
        manufacturer: Intel
        model: Intel
        serial-number: 0000000000000000
        hardware-version: 1.16
        firmware-version: 45.9
        flags:
               rng
               login-required
               user-pin-initialized
               token-initialized
++ p11tool --list-token-urls
++ grep Intel
+ TOKEN='pkcs11:model=Intel;manufacturer=Intel;serial=0000000000000000;token=label'
+ expect /dev/fd/63
++ cat
spawn p11tool --login --list-all pkcs11:model=Intel;manufacturer=Intel;serial=0000000000000000;token=label
Token 'label' with URL 'pkcs11:model=Intel;manufacturer=Intel;serial=0000000000000000;token=label' requires user PIN
Enter PIN: 
Object 0:
        URL: pkcs11:model=Intel;manufacturer=Intel;serial=0000000000000000;token=label;id=%33%38%64%36%32%64%37%65%33%65%64%61%38%34%65%62;type=private
        Type: Private key (RSA-2048)
        Label: 
        Flags: CKA_PRIVATE; CKA_NEVER_EXTRACTABLE; CKA_SENSITIVE; 
        ID: 33:38:64:36:32:64:37:65:33:65:64:61:38:34:65:62

Object 1:
        URL: pkcs11:model=Intel;manufacturer=Intel;serial=0000000000000000;token=label;id=%33%38%64%36%32%64%37%65%33%65%64%61%38%34%65%62;type=public
WARNING:esys:src/tss2-esys/api/Esys_StartAuthSession.c:383:Esys_StartAuthSession_Finish() Received TPM Error 
ERROR:esys:src/tss2-esys/api/Esys_StartAuthSession.c:136:Esys_StartAuthSession() Esys Finish ErrorCode (0x00000902) 
ERROR: Esys_StartAuthSession: 0x902
        Type: Public key
        Label: 
        ID: 33:38:64:36:32:64:37:65:33:65:64:61:38:34:65:62
```

Is this a tpm2-tss, tpm2-tools, tpm2-pkcs11 or p11tool issue?

**Comment by williamcroberts** [comment]
Can you `export TPM2_PKCS11_LOG_LEVEL=2` and post that log. IIRC ive seen this before, I think that's calling C_Login() on an unitialized token.

**Comment by dnoliver** [comment]
Output with `export TPM2_PKCS11_LOG_LEVE=2`

```text
+ export TPM2TOOLS_TCTI=device:/dev/tpmrm0
+ TPM2TOOLS_TCTI=device:/dev/tpmrm0
+ export TPM2_PKCS11_LOG_LEVEL=2
+ TPM2_PKCS11_LOG_LEVEL=2
+ tpm2_clear
+ rm tpm2_pkcs11.sqlite3
+ tpm2_ptool init
action: Created
id: 1

+ tpm2_ptool addtoken --pid=1 --sopin=mysopin --userpin=myuserpin --label=label
+ tpm2_ptool addkey --algorithm=rsa2048 --label=label --userpin=myuserpin
action: add
private:
  CKA_ID: '38653330306164303132633139333539'
public:
  CKA_ID: '38653330306164303132633139333539'
+ p11-kit list-modules
INFO on line: "387" in file: "src/pkcs11.c": enter "C_GetFunctionList"
INFO on line: "387" in file: "src/pkcs11.c": return "C_GetFunctionList" value: 0
INFO on line: "375" in file: "src/pkcs11.c": enter "C_Initialize"
INFO on line: "909" in file: "src/lib/db.c": Could not stat db at path "/root/.tpm2_pkcs11/tpm2_pkcs11.sqlite3", error: No such file or directory
INFO on line: "941" in file: "src/lib/db.c": Using sqlite3 DB: "/root/tpm2_pkcs11.sqlite3"
INFO on line: "267" in file: "src/lib/tpm.c": tcti=(null)
INFO on line: "375" in file: "src/pkcs11.c": return "C_Initialize" value: 0
p11-kit-trust: p11-kit-trust.so
    library-description: PKCS#11 Kit Trust Module
    library-manufacturer: PKCS#11 Kit
    library-version: 0.23
    token: System Trust
        manufacturer: PKCS#11 Kit
        model: p11-kit-trust
        serial-number: 1
        hardware-version: 0.23
        flags:
               token-initialized
    token: Default Trust
        manufacturer: PKCS#11 Kit
        model: p11-kit-trust
        serial-number: 1
        hardware-version: 0.23
        flags:
               write-protected
               token-initialized
opensc: opensc-pkcs11.so
    library-description: OpenSC smartcard framework
    library-manufacturer: OpenSC Project
    library-version: 0.19
tpm2_pkcs11: libtpm2_pkcs11.so
INFO on line: "383" in file: "src/pkcs11.c": enter "C_GetInfo"
INFO on line: "383" in file: "src/pkcs11.c": return "C_GetInfo" value: 0
    library-description: TPM2.0 Cryptoki
    library-manufacturer: tpm2-software.github.io
    library-version: 1.0
INFO on line: "391" in file: "src/pkcs11.c": enter "C_GetSlotList"
INFO on line: "391" in file: "src/pkcs11.c": return "C_GetSlotList" value: 0
INFO on line: "399" in file: "src/pkcs11.c": enter "C_GetTokenInfo"
INFO on line: "399" in file: "src/pkcs11.c": return "C_GetTokenInfo" value: 0
    token: label
        manufacturer: Intel
        model: Intel
        serial-number: 0000000000000000
        hardware-version: 1.16
        firmware-version: 45.9
        flags:
               rng
               login-required
               user-pin-initialized
               token-initialized
INFO on line: "379" in file: "src/pkcs11.c": enter "C_Finalize"
INFO on line: "379" in file: "src/pkcs11.c": return "C_Finalize" value: 0
++ p11tool --list-token-urls
++ grep Intel
INFO on line: "387" in file: "src/pkcs11.c": enter "C_GetFunctionList"
INFO on line: "387" in file: "src/pkcs11.c": return "C_GetFunctionList" value: 0
INFO on line: "375" in file: "src/pkcs11.c": enter "C_Initialize"
INFO on line: "909" in file: "src/lib/db.c": Could not stat db at path "/root/.tpm2_pkcs11/tpm2_pkcs11.sqlite3", error: No such file or directory
INFO on line: "941" in file: "src/lib/db.c": Using sqlite3 DB: "/root/tpm2_pkcs11.sqlite3"
INFO on line: "267" in file: "src/lib/tpm.c": tcti=(null)
INFO on line: "375" in file: "src/pkcs11.c": return "C_Initialize" value: 0
INFO on line: "383" in file: "src/pkcs11.c": enter "C_GetInfo"
INFO on line: "383" in file: "src/pkcs11.c": return "C_GetInfo" value: 0
INFO on line: "391" in file: "src/pkcs11.c": enter "C_GetSlotList"
INFO on line: "391" in file: "src/pkcs11.c": return "C_GetSlotList" value: 0
INFO on line: "399" in file: "src/pkcs11.c": enter "C_GetTokenInfo"
INFO on line: "399" in file: "src/pkcs11.c": return "C_GetTokenInfo" value: 0
INFO on line: "395" in file: "src/pkcs11.c": enter "C_GetSlotInfo"
INFO on line: "395" in file: "src/pkcs11.c": return "C_GetSlotInfo" value: 0
INFO on line: "427" in file: "src/pkcs11.c": enter "C_OpenSession"
INFO on line: "427" in file: "src/pkcs11.c": return "C_OpenSession" value: 0
INFO on line: "431" in file: "src/pkcs11.c": enter "C_CloseSession"
INFO on line: "431" in file: "src/pkcs11.c": return "C_CloseSession" value: 0
INFO on line: "391" in file: "src/pkcs11.c": enter "C_GetSlotList"
INFO on line: "391" in file: "src/pkcs11.c": return "C_GetSlotList" value: 0
INFO on line: "399" in file: "src/pkcs11.c": enter "C_GetTokenInfo"
INFO on line: "399" in file: "src/pkcs11.c": return "C_GetTokenInfo" value: 0
INFO on line: "395" in file: "src/pkcs11.c": enter "C_GetSlotInfo"
INFO on line: "395" in file: "src/pkcs11.c": return "C_GetSlotInfo" value: 0
INFO on line: "427" in file: "src/pkcs11.c": enter "C_OpenSession"
INFO on line: "427" in file: "src/pkcs11.c": return "C_OpenSession" value: 0
INFO on line: "431" in file: "src/pkcs11.c": enter "C_CloseSession"
INFO on line: "431" in file: "src/pkcs11.c": return "C_CloseSession" value: 0
INFO on line: "379" in file: "src/pkcs11.c": enter "C_Finalize"
INFO on line: "379" in file: "src/pkcs11.c": return "C_Finalize" value: 0
+ TOKEN='pkcs11:model=Intel;manufacturer=Intel;serial=0000000000000000;token=label'
+ expect /dev/fd/63
++ cat
spawn p11tool --login --list-all pkcs11:model=Intel;manufacturer=Intel;serial=0000000000000000;token=label --outfile p11tool.out
INFO on line: "387" in file: "src/pkcs11.c": enter "C_GetFunctionList"
INFO on line: "387" in file: "src/pkcs11.c": return "C_GetFunctionList" value: 0
INFO on line: "375" in file: "src/pkcs11.c": enter "C_Initialize"
INFO on line: "909" in file: "src/lib/db.c": Could not stat db at path "/root/.tpm2_pkcs11/tpm2_pkcs11.sqlite3", error: No such file or directory
INFO on line: "941" in file: "src/lib/db.c": Using sqlite3 DB: "/root/tpm2_pkcs11.sqlite3"
INFO on line: "267" in file: "src/lib/tpm.c": tcti=(null)
INFO on line: "375" in file: "src/pkcs11.c": return "C_Initialize" value: 0
INFO on line: "383" in file: "src/pkcs11.c": enter "C_GetInfo"
INFO on line: "383" in file: "src/pkcs11.c": return "C_GetInfo" value: 0
INFO on line: "391" in file: "src/pkcs11.c": enter "C_GetSlotList"
INFO on line: "391" in file: "src/pkcs11.c": return "C_GetSlotList" value: 0
INFO on line: "399" in file: "src/pkcs11.c": enter "C_GetTokenInfo"
INFO on line: "399" in file: "src/pkcs11.c": return "C_GetTokenInfo" value: 0
INFO on line: "395" in file: "src/pkcs11.c": enter "C_GetSlotInfo"
INFO on line: "395" in file: "src/pkcs11.c": return "C_GetSlotInfo" value: 0
INFO on line: "427" in file: "src/pkcs11.c": enter "C_OpenSession"
INFO on line: "427" in file: "src/pkcs11.c": return "C_OpenSession" value: 0
INFO on line: "431" in file: "src/pkcs11.c": enter "C_CloseSession"
INFO on line: "431" in file: "src/pkcs11.c": return "C_CloseSession" value: 0
INFO on line: "391" in file: "src/pkcs11.c": enter "C_GetSlotList"
INFO on line: "391" in file: "src/pkcs11.c": return "C_GetSlotList" value: 0
INFO on line: "399" in file: "src/pkcs11.c": enter "C_GetTokenInfo"
INFO on line: "399" in file: "src/pkcs11.c": return "C_GetTokenInfo" value: 0
INFO on line: "395" in file: "src/pkcs11.c": enter "C_GetSlotInfo"
INFO on line: "395" in file: "src/pkcs11.c": return "C_GetSlotInfo" value: 0
INFO on line: "427" in file: "src/pkcs11.c": enter "C_OpenSession"
INFO on line: "427" in file: "src/pkcs11.c": return "C_OpenSession" value: 0
INFO on line: "439" in file: "src/pkcs11.c": enter "C_GetSessionInfo"
INFO on line: "439" in file: "src/pkcs11.c": return "C_GetSessionInfo" value: 0
Token 'label' with URL 'pkcs11:model=Intel;manufacturer=Intel;serial=0000000000000000;token=label' requires user PIN
Enter PIN: 
INFO on line: "451" in file: "src/pkcs11.c": enter "C_Login"
INFO on line: "451" in file: "src/pkcs11.c": return "C_Login" value: 0
INFO on line: "483" in file: "src/pkcs11.c": enter "C_FindObjectsInit"
INFO on line: "483" in file: "src/pkcs11.c": return "C_FindObjectsInit" value: 0
INFO on line: "487" in file: "src/pkcs11.c": enter "C_FindObjects"
INFO on line: "487" in file: "src/pkcs11.c": return "C_FindObjects" value: 0
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 0
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 0
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "400" in file: "src/lib/object.c": Invalid Attribute for tid 1: type(262) ulValueLen(18446744073709551615), pData(null)
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 18
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 0
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 0
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "400" in file: "src/lib/object.c": Invalid Attribute for tid 1: type(134) ulValueLen(18446744073709551615), pData(null)
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 18
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 0
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 0
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 0
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "400" in file: "src/lib/object.c": Invalid Attribute for tid 1: type(135) ulValueLen(18446744073709551615), pData(null)
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 18
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 0
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 0
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 0
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "400" in file: "src/lib/object.c": Invalid Attribute for tid 1: type(17) ulValueLen(18446744073709551615), pData(null)
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 18
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 0
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 0
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 0
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "400" in file: "src/lib/object.c": Invalid Attribute for tid 2: type(263) ulValueLen(18446744073709551615), pData(null)
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 18
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 0
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 0
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "400" in file: "src/lib/object.c": Invalid Attribute for tid 2: type(259) ulValueLen(18446744073709551615), pData(null)
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 18
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "400" in file: "src/lib/object.c": Invalid Attribute for tid 2: type(354) ulValueLen(18446744073709551615), pData(null)
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 18
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "400" in file: "src/lib/object.c": Invalid Attribute for tid 2: type(356) ulValueLen(18446744073709551615), pData(null)
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 18
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "400" in file: "src/lib/object.c": Invalid Attribute for tid 2: type(135) ulValueLen(18446744073709551615), pData(null)
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 18
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "400" in file: "src/lib/object.c": Invalid Attribute for tid 2: type(514) ulValueLen(18446744073709551615), pData(null)
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 18
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 0
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 0
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "400" in file: "src/lib/object.c": Invalid Attribute for tid 2: type(17) ulValueLen(18446744073709551615), pData(null)
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 18
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 0
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 0
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 0
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 0
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 0
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 0
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 0
INFO on line: "487" in file: "src/pkcs11.c": enter "C_FindObjects"
INFO on line: "487" in file: "src/pkcs11.c": return "C_FindObjects" value: 0
INFO on line: "491" in file: "src/pkcs11.c": enter "C_FindObjectsFinal"
INFO on line: "491" in file: "src/pkcs11.c": return "C_FindObjectsFinal" value: 0
INFO on line: "431" in file: "src/pkcs11.c": enter "C_CloseSession"
INFO on line: "431" in file: "src/pkcs11.c": return "C_CloseSession" value: 0
INFO on line: "391" in file: "src/pkcs11.c": enter "C_GetSlotList"
INFO on line: "391" in file: "src/pkcs11.c": return "C_GetSlotList" value: 0
INFO on line: "399" in file: "src/pkcs11.c": enter "C_GetTokenInfo"
INFO on line: "399" in file: "src/pkcs11.c": return "C_GetTokenInfo" value: 0
INFO on line: "395" in file: "src/pkcs11.c": enter "C_GetSlotInfo"
INFO on line: "395" in file: "src/pkcs11.c": return "C_GetSlotInfo" value: 0
INFO on line: "427" in file: "src/pkcs11.c": enter "C_OpenSession"
INFO on line: "427" in file: "src/pkcs11.c": return "C_OpenSession" value: 0
INFO on line: "439" in file: "src/pkcs11.c": enter "C_GetSessionInfo"
INFO on line: "439" in file: "src/pkcs11.c": return "C_GetSessionInfo" value: 0
INFO on line: "451" in file: "src/pkcs11.c": enter "C_Login"
INFO on line: "451" in file: "src/pkcs11.c": return "C_Login" value: 0
INFO on line: "483" in file: "src/pkcs11.c": enter "C_FindObjectsInit"
INFO on line: "483" in file: "src/pkcs11.c": return "C_FindObjectsInit" value: 0
INFO on line: "487" in file: "src/pkcs11.c": enter "C_FindObjects"
INFO on line: "487" in file: "src/pkcs11.c": return "C_FindObjects" value: 0
INFO on line: "491" in file: "src/pkcs11.c": enter "C_FindObjectsFinal"
INFO on line: "491" in file: "src/pkcs11.c": return "C_FindObjectsFinal" value: 0
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 0
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 0
INFO on line: "391" in file: "src/pkcs11.c": enter "C_GetSlotList"
INFO on line: "391" in file: "src/pkcs11.c": return "C_GetSlotList" value: 0
INFO on line: "399" in file: "src/pkcs11.c": enter "C_GetTokenInfo"
INFO on line: "399" in file: "src/pkcs11.c": return "C_GetTokenInfo" value: 0
INFO on line: "395" in file: "src/pkcs11.c": enter "C_GetSlotInfo"
INFO on line: "395" in file: "src/pkcs11.c": return "C_GetSlotInfo" value: 0
INFO on line: "411" in file: "src/pkcs11.c": enter "C_GetMechanismInfo"
INFO on line: "411" in file: "src/pkcs11.c": return "C_GetMechanismInfo" value: 112
INFO on line: "475" in file: "src/pkcs11.c": enter "C_GetAttributeValue"
INFO on line: "475" in file: "src/pkcs11.c": return "C_GetAttributeValue" value: 0
INFO on line: "431" in file: "src/pkcs11.c": enter "C_CloseSession"
INFO on line: "431" in file: "src/pkcs11.c": return "C_CloseSession" value: 0
INFO on line: "391" in file: "src/pkcs11.c": enter "C_GetSlotList"
INFO on line: "391" in file: "src/pkcs11.c": return "C_GetSlotList" value: 0
INFO on line: "399" in file: "src/pkcs11.c": enter "C_GetTokenInfo"
INFO on line: "399" in file: "src/pkcs11.c": return "C_GetTokenInfo" value: 0
INFO on line: "395" in file: "src/pkcs11.c": enter "C_GetSlotInfo"
INFO on line: "395" in file: "src/pkcs11.c": return "C_GetSlotInfo" value: 0
INFO on line: "427" in file: "src/pkcs11.c": enter "C_OpenSession"
INFO on line: "427" in file: "src/pkcs11.c": return "C_OpenSession" value: 0
INFO on line: "439" in file: "src/pkcs11.c": enter "C_GetSessionInfo"
INFO on line: "439" in file: "src/pkcs11.c": return "C_GetSessionInfo" value: 0
INFO on line: "451" in file: "src/pkcs11.c": enter "C_Login"
WARNING:esys:src/tss2-esys/api/Esys_StartAuthSession.c:383:Esys_StartAuthSession_Finish() Received TPM Error 
ERROR:esys:src/tss2-esys/api/Esys_StartAuthSession.c:136:Esys_StartAuthSession() Esys Finish ErrorCode (0x00000902) 
ERROR on line: "216" in file: "src/lib/tpm.c": Esys_StartAuthSession: 0x902
INFO on line: "451" in file: "src/pkcs11.c": return "C_Login" value: 5
INFO on line: "431" in file: "src/pkcs11.c": enter "C_CloseSession"
INFO on line: "431" in file: "src/pkcs11.c": return "C_CloseSession" value: 0
INFO on line: "379" in file: "src/pkcs11.c": enter "C_Finalize"
INFO on line: "379" in file: "src/pkcs11.c": return "C_Finalize" value: 0
```

**Comment by williamcroberts** [comment]
Have you tried p11tool with both TCTI env variables set. tpm2_ptool calls to the tools and p11tool should be respecting the pkcs11 config. Are there by any chance sessions not flushed from the TPM, so your starting in a state with sessions not available?

**Comment by dnoliver** [comment]
Yes, I tried bot env vars and tpm2_ptool config and I always get that error.
I also took into account cleaning every handle manually. I am trying to get this reproducible, so also clearing tpm and deleting the pkcs11 database.

**Comment by williamcroberts** [comment]
I'd also through in some tpm2_getcap calls to get the state of the TPM, maybe we can figure out exactly where the state is going awry.
```bash
for p in handles-loaded-session handles-saved-session handles-transient; do
  echo "---$p---"
  tpm2_getcap "$p"
  echo "------";
done;
```

**Comment by dnoliver** [comment]
I re-validated this in master branch commit 9c77c565413d6760700d6622e523f96686146e0d and it is no longer reproducible :). ptool does not complains.

I think my problem is with this line in the script

`TOKEN=$(p11tool --list-token-urls | grep "Intel")`

With some versions of tpm2_pkcs11, I get 2 tokens in my pkcs11 store. One has an empty label "manufacturer=Intel;token=", the other one has the label that I selected. The original version I was testing only listed one, so I thought it was safe to select the token based on manufacturer.

So, when calling p11tool, there were 2 tokens passed to the `--list-all` parameter, and I got the credentials for both of them asked (the real one with an "Enter PIN" label, the bad one with an "Enter password" label). The password for the other token was not correct, so p11tool printed that error.

This fixed it:

`TOKEN=$(p11tool --list-token-urls | grep "token=label")`

I have no idea why I sometimes end up with that empty label token:

```bash
[root@localhost ~]# p11tool --list-token-urls
pkcs11:model=p11-kit-trust;manufacturer=PKCS%2311%20Kit;serial=1;token=System%20Trust
pkcs11:model=p11-kit-trust;manufacturer=PKCS%2311%20Kit;serial=1;token=Default%20Trust
pkcs11:model=Intel;manufacturer=Intel;serial=0000000000000000;token=label
pkcs11:model=Intel;manufacturer=Intel;serial=0000000000000000;token=
```

anyway, my bad!
